src/crimson/os/seastore/extentmap_manager/btree/extentmap_btree_node_impl.cc

```diff
@@ -0,0 +1,604 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include <sys/mman.h>
+#include <string.h>
+
+#include <memory>
+#include <string.h>
+
+#include "include/buffer.h"
+#include "include/byteorder.h"
+#include "crimson/os/seastore/transaction_manager.h"
+#include "crimson/os/seastore/extentmap_manager/btree/extentmap_btree_node.h"
+#include "crimson/os/seastore/extentmap_manager/btree/extentmap_btree_node_impl.h"
+
+#include "test/crimson/seastore/test_block.h"   //for temporary
+namespace {
+  seastar::logger& logger() {
+    return crimson::get_logger(ceph_subsys_filestore);
+  }
+}
+
+namespace crimson::os::seastore::extentmap_manager {
+
+std::ostream &ExtMapInnerNode::print_detail(std::ostream &out) const
+{
+  return out << ", size=" << get_size()
+	     << ", depth=" << depth;
+}
+ExtMapInnerNode::seek_lextent_ret
+ExtMapInnerNode::seek_lextent(Transaction &t, objaddr_t lo, extent_len_t len)
+{
+  auto [begin, end] = bound(lo, lo + len);
+  auto result_up = std::make_unique<extent_map_list_t>();
+  auto &result = *result_up;
+  return crimson::do_for_each(
+    std::move(begin),
+    std::move(end),
+    [this, &t, &result, lo, len](const auto &val) mutable {
+      return extmap_load_extent(tm, t, val.get_val(), depth - 1).safe_then(
+	  [&t, &result, lo, len](auto extent) mutable {
+	    // TODO: add backrefs to ensure cache residence of parents
+	    return extent->seek_lextent(t, lo, len).safe_then(
+		[&t, &result, lo, len](auto item_list) mutable {
+		  result.splice(result.end(), item_list,
+				item_list.begin(), item_list.end());
+            });
+      });
+  }).safe_then([result=std::move(result_up)] {
+    return seek_lextent_ret(
+           seek_lextent_ertr::ready_future_marker{},
+           std::move(*result));
+  });
+}
+
+ExtMapInnerNode::insert_ret
+ExtMapInnerNode::insert(Transaction &t, objaddr_t lo, lext_map_val_t val)
+{
+  auto insertion_pt = get_containing_child(lo);
```

Comment:
assert not end

Reply:
could you explain it a little?

Reply:
You are assuming that insertion_pt != end().  Add an assert to that effect.